Enhanced 3D model support.

IceReaper opened the pull request:
This PR implement a bunch of 3d model api changes:

- First any mod can now define supported 3d formats like sprite formats.
- As every model format might require different rechniques, a model now defined its VertexBuffer and Shader.

A new IShaderBinding interface has been created to properly configure the Shader class, without the need to make the OpenGL namespace accessable in the mods.

Here is a simple example where i load .obj models, and apply a custom IShaderBinding along  with a custom Vertex struct and a new glsl shader which makes use of a RGBA DiffuseMap as well as vertex normals:

![Animation](https://user-images.githubusercontent.com/1322277/122783657-0a139800-d2b2-11eb-8135-7ecc0fdcb8f1.gif)

For my prototype, i also implemented a trait which glues submodels onto bones: Animated legs model with abone for another animated torso model. And that has 2 bones for the head and weapon models.

![Animation](https://user-images.githubusercontent.com/1322277/123374246-17b57000-d587-11eb-8834-1e0bc71b2c75.gif)

Depends on #19519
Closes #4689

abcdefg30 commented on OpenRA.Game/Graphics/ModelCache.cs:
```diff
@@ -0,0 +1,80 @@
+﻿#region Copyright & License Information
+/*
+ * Copyright 2007-2020 The OpenRA Developers (see AUTHORS)
```
2021

abcdefg30 commented on OpenRA.Game/Graphics/ShaderBindings.cs:
```diff
@@ -0,0 +1,60 @@
+﻿#region Copyright & License Information
+
+/*
+ * Copyright 2007-2020 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+
+#endregion
```
```suggestion
﻿#region Copyright & License Information
/*
 * Copyright 2007-2021 The OpenRA Developers (see AUTHORS)
 * This file is part of OpenRA, which is free software. It is made
 * available to you under the terms of the GNU General Public License
 * as published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version. For more
 * information, see COPYING.
 */
#endregion
```

abcdefg30 commented on OpenRA.Game/Graphics/ShaderVertexAttribute.cs:
```diff
@@ -0,0 +1,29 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2020 The OpenRA Developers (see AUTHORS)
```
2021

teinarss commented on OpenRA.Platforms.Default/ThreadedGraphicsContext.cs:
```diff
@@ -25,7 +26,7 @@ namespace OpenRA.Platforms.Default
 	{
 		// PERF: Maintain several object pools to reduce allocations.
-		readonly Stack<Vertex[]> verticesPool = new Stack<Vertex[]>();
+		readonly Stack<object[]> verticesPool = new Stack<object[]>();
```
This will cause boxing, without doing any perf testing we cant say for sure what performance implications this will cause.

pchote commented:
This is currently doing too many distinct things for one PR (let alone one commit!).

I haven't looked at this in detail to look for any show stoppers yet, but if we do want to move forward with these changes we should split the distinct features (e.g. at least custom shader support, renaming voxel &rarr; model in traits etc, implementing general model loaders) into distinct PRs.

IceReaper commented:
Ill split the pr into several smaller ones as soon as i can confirm gltf support works as promised and above PR doesnt need changes to allow it. I made the PR pretty early, so one might quickly oversee it and throw in enhancements or wishes early.

IceReaper committed:
Implemented ModelFormats.

IceReaper committed:
Replaced hardcoded usage of voxels with models.

IceReaper committed:
Implement mod defined model shaders.

IceReaper committed:
Support for bones and transforms (to glue models by bones).

IceReaper commented:
As i have successfully used 3d models now for a while, including:
- GLTF model loading
- Defining my own shader to properly render models as desired.
- Shared animations for multiple models (1 walk animation for X character models)
- Using bones to assemble actors from multiple models
- Removed the need of the SheetBuilder to speed up the loading times for model textures by almost 500%

I refactored this pr, and moved all the changes into individual commits.
So i consider this PR to be ready to review :)

pchote commented:
I unfortunately don't have the time nowadays to sit down for a full day (or more!) to get my head around complicated PRs 😢 

This PR is doing too much for me to even start to review. Can we please start by splitting the custom shader support out into its own PR? That has important interactions with other changes that we need in the near future for remaster, #9778, etc support, so I would like to be able to focus on getting that right first without distractions from all the other changes here.

lukebrowell commented:
@pchote this seems worthy - how can we help with the review?

pchote commented:
This is blocked on #19519.

dnqbob commented:
Hi man, would you mind helping us on adding those change into our engine? We are dying on making soldier SHP to support our infantry system (our infantry have animation like attack->move, attack->idle animation which take space and take unbearable time to render into sequences), as well as it is impossible to make airborne sequence on pitch and roll.

abcdefg30 commented:
Closing as superseded by https://github.com/OpenRA/OpenRA/pull/20581.
